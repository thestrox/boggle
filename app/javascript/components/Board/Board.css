.board {
    background-color: lightblue;
}

.board-row {
    display: flex;
    flex-direction: row;
}