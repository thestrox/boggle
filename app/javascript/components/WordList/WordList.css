.wordlist {
    border: 1px solid lightgray;
    padding-top: 0 !important;
    padding-bottom: 0 !important;
    overflow: auto;
}

.wordlist-item {
    justify-content: space-between !important;
}