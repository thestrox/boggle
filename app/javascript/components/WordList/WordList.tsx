import React from "react";

export type WordListProp = {
    wordScoreMap: { [key: string]: number }
}

export default function WordList({ wordScoreMap }: WordListProp) {

    return (
        <ul>
            {Object.keys(wordScoreMap).map(word => {
                return (
                    <li key={word}>{word} : {wordScoreMap[word]}</li>
                );
            })}
        </ul>
    )
}