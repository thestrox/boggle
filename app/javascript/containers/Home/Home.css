.home-container {
    display: flex;
    flex-direction: row;
}

.left-container {
    margin-right: 50px;
}

.board-container, .wordlist-container  {
    min-height: 200px;
}

.wordlist-container {
    height: 250px;
}

.word-input {
    margin-bottom: 10px;
}

.reset-button, .end-button, .submit-button, .new-button {
    margin-right: 10px !important;
}

.button-wrapper {
    margin-top: 10px !important;
}