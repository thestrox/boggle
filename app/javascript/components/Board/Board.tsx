import React from "react";
import Tile from "../Tile/Tile";
import "./Board.css"

export type BoardProps = {
    board: string[][],
    className: string,
    onTileClick: (row: number, col: number) => void
}

export default function Board({ board, className, onTileClick }: BoardProps) {

    return (
        <div className={`board ${className}`}>
            {board.map((row, rowIndex) => {
                return (
                    <div className="board-row" key={rowIndex}>
                        {row.map((letter, colIndex) => {
                            return (<Tile key={colIndex} onClick={onTileClick.bind(this, rowIndex, colIndex)} letter={letter} />);
                        })}
                    </div>
                );
            })}
        </div >
    );

}