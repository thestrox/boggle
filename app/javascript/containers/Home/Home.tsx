import React from "react";
import Board from "../../components/Board/Board";
import { connect } from "react-redux";
import { AppState, AppThunk } from "../../store";
import { initializeNewBoard, validateWord, resetBoard } from "../../store/thunks";
import WordList from "../../components/WordList/WordList";
import { ValidateRequest } from "../../shared/types/api-types";
import { toast } from "react-toastify";
import Countdown, { zeroPad } from "react-countdown";
import { Redirect } from "react-router-dom";
import Button from '@material-ui/core/Button';
import "./Home.css"
import { Input } from "@material-ui/core";

export type HomeProps = {
    board: string[][];
    wordScoreMap: { [key: string]: number }
    duration: number;
    gameStartDate: number;
    initializeNewBoard: () => AppThunk,
    validateWord: (validateRequestBody: ValidateRequest) => AppThunk
    resetBoard: () => AppThunk
}

export type HomeState = {
    currentWord: string;
    toResult: boolean;
};

class Home extends React.Component<HomeProps, HomeState> {

    constructor(props: HomeProps) {
        super(props);
        this.state = {
            currentWord: '',
            toResult: false
        };
    }

    componentDidMount() {
        this.props.initializeNewBoard();
    }

    reset = () => {
        this.props.resetBoard();
    }

    newGame = () => {
        this.props.initializeNewBoard();
    }

    endGame = () => {
        this.setState({
            toResult: true
        });
    }

    validateWord = () => {
        if (this.props.wordScoreMap[this.state.currentWord]) {
            toast.error('Duplicate word. Please try again.');
        } else {
            this.props.validateWord({ board: this.props.board, word: this.state.currentWord });
        }
        this.updateCurrentWord("");
    }

    onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        this.updateCurrentWord(event.target.value);
    }

    updateCurrentWord = (word: string) => {
        this.setState({ currentWord: word.toUpperCase() });
    }

    keyPress = (e: React.KeyboardEvent<any>) => {
        if (e.key === "Enter") {
            this.validateWord();
        }
    }

    onTileClick = (row: number, col: number) => {
        this.updateCurrentWord(this.state.currentWord + this.props.board[row][col]);
    }

    render() {
        if (this.state.toResult) {
            return (
                <Redirect to='/result' />
            );
        }

        return (
            <div className="home-container">
                <div className="left-container">
                    <Input className="word-input"
                        value={this.state.currentWord}
                        onChange={this.onChange}
                        onKeyPress={this.keyPress}
                        placeholder="Type a suitable word"
                    />

                    <Board board={this.props.board} onTileClick={this.onTileClick} className="board-container"></Board>
                    <div className="button-wrapper">
                        <Button classes={{ root: 'submit-button' }} variant="contained" color="primary" onClick={this.validateWord}>Submit Word</Button>
                        <Button classes={{ root: 'end-button' }} variant="contained" color="default" onClick={this.endGame}>End Game</Button>
                    </div>

                </div>
                <div className="right-container">
                    {this.props.gameStartDate &&
                        <Countdown
                            date={this.props.gameStartDate + this.props.duration}
                            onComplete={this.endGame}
                            renderer={(props) => <span>{zeroPad(props.minutes)}:{zeroPad(props.seconds)}</span>}
                        />
                    }
                    <div>Results: </div>
                    <WordList wordScoreMap={this.props.wordScoreMap} className="wordlist-container"></WordList>
                    <div className="button-wrapper">
                        <Button classes={{ root: 'reset-button' }} variant="contained" onClick={this.reset}>Reset</Button>
                        <Button classes={{ root: 'new-button' }} variant="contained" onClick={this.newGame}>New Game</Button>
                    </div>


                </div>
            </div >
        );
    }
}

const mapStateToProps = (state: AppState) => {
    return {
        board: state.board.board,
        wordScoreMap: state.score.wordScoreMap,
        duration: state.board.duration,
        gameStartDate: state.board.gameStartDate,
    }
}

const HomeConnected = connect(
    mapStateToProps,
    {
        initializeNewBoard: initializeNewBoard,
        validateWord: validateWord,
        resetBoard: resetBoard
    }
)(Home)
export default HomeConnected