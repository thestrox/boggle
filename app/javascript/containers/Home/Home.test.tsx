import React from 'react';
import { shallow } from 'enzyme';
import {Home} from "./Home"
import Board from '../../components/Board/Board';
import WordList from '../../components/WordList/WordList';
import Countdown from 'react-countdown';
import { Input, Button } from '@material-ui/core';

function setup() {
    const props = {
        board: [
            ["A", "B", "C", "D"],
            ["E", "F", "G", "H"],
            ["I", "J", "K", "L"],
            ["M", "N", "O", "P"]
        ],
        wordScoreMap: {ONE: 3},
        duration: 1000,
        gameStartDate: 1234,
        initializeNewBoard: jest.fn(),
        validateWord: jest.fn(),
        resetBoard: jest.fn()
    };

    const component = shallow<Home>(<Home {...props}/>);
    return {
        props,
        component
    }
}

describe('Home Component', () => {
    it('should render correctly with given props', () => {
        const {component, props} = setup();
        expect(component.find(Board).prop('board')).toBe(props.board);
        expect(component.find(Countdown)).toBeTruthy();
        expect(component.find(Countdown).prop('date')).toBe(2234);
        expect(component.find(WordList).prop('wordScoreMap')).toBe(props.wordScoreMap);
    });

    it('onChange: should update the state.currentWord', () => {
        const {component} = setup();
        const instance = component.instance();
        expect(instance.state.currentWord).toBe('');
        const value = 'NEWWORD';
        const mockedEvent = {target: {value}}
        component.find(Input).simulate('change', mockedEvent);
        expect(instance.state.currentWord).toBe(value);
    
    });

    it('onKeyPress: should call validateWord on "Enter"', () => {
        const {component, props} = setup();
        const instance = component.instance();
        instance.updateCurrentWord('tone');
        component.find(Input).simulate('keypress', {key: 'Enter'});
        expect(props.validateWord).toHaveBeenCalled();
    });

    it('onTileClick: should update the state.currentWord', () => {
        const {component, props} = setup();
        const instance = component.instance();
        expect(instance.state.currentWord).toBe('');
        component.find(Board).props().onTileClick(0,0);
        expect(instance.state.currentWord).toBe(props.board[0][0].toUpperCase());

    })

    it('onSubmitClick: should call props validateWord', () => {
        const {component, props} = setup();
        const instance = component.instance();
        instance.updateCurrentWord('tone');
        component.find(Button).at(0).simulate('click');
        expect(props.validateWord).toHaveBeenCalled();
    })

    it('onSubmitClick: should not call props validateWord', () => {
        const {component, props} = setup();
        const instance = component.instance();
        instance.updateCurrentWord('one');
        component.find(Button).at(0).simulate('click');
        expect(props.validateWord).toHaveBeenCalledTimes(0);
    })

    it('onEndGameClick: should call change state.toResult to true', () => {
        const {component} = setup();
        const instance = component.instance();
        expect(instance.state.toResult).toBe(false);
        component.find(Button).at(1).simulate('click');
        expect(instance.state.toResult).toBe(true);
    })

    it('should call change state.toResult to true on Countdown complete', () => {
        const {component} = setup();
        const instance = component.instance();
        expect(instance.state.toResult).toBe(false);
        component.find(Countdown).props().onComplete(null);
        expect(instance.state.toResult).toBe(true);
    })

    it('onResetClick: should call resetBoard', () => {
        const {component, props} = setup();
        component.find(Button).at(2).simulate('click');
        expect(props.resetBoard).toHaveBeenCalled();
    })

    it('onNewGameClick: should call initializeNewBoard', () => {
        const {component, props} = setup();
        component.find(Button).at(3).simulate('click');
        expect(props.initializeNewBoard).toHaveBeenCalled();
    })

    it('should test snapshot renders', () => {
        const {component} = setup();
        expect(component).toMatchSnapshot();
    });
})