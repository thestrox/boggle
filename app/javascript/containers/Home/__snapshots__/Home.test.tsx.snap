// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should test snapshot renders 1`] = `
<div
  className="home-container"
>
  <div
    className="left-container"
  >
    <WithStyles(ForwardRef(Input))
      className="word-input"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type a suitable word"
      value=""
    />
    <Board
      board={
        Array [
          Array [
            "A",
            "B",
            "C",
            "D",
          ],
          Array [
            "E",
            "F",
            "G",
            "H",
          ],
          Array [
            "I",
            "J",
            "K",
            "L",
          ],
          Array [
            "M",
            "N",
            "O",
            "P",
          ],
        ]
      }
      className="board-container"
      onTileClick={[Function]}
    />
    <div
      className="button-wrapper"
    >
      <WithStyles(ForwardRef(Button))
        classes={
          Object {
            "root": "submit-button",
          }
        }
        color="primary"
        onClick={[Function]}
        variant="contained"
      >
        Submit Word
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        classes={
          Object {
            "root": "end-button",
          }
        }
        color="default"
        onClick={[Function]}
        variant="contained"
      >
        End Game
      </WithStyles(ForwardRef(Button))>
    </div>
  </div>
  <div
    className="right-container"
  >
    <Countdown$1
      autoStart={true}
      controlled={false}
      date={2234}
      daysInHours={false}
      intervalDelay={1000}
      onComplete={[Function]}
      precision={0}
      renderer={[Function]}
      zeroPadTime={2}
    />
    <div>
      Results: 
    </div>
    <WordList
      className="wordlist-container"
      wordScoreMap={
        Object {
          "ONE": 3,
        }
      }
    />
    <div
      className="button-wrapper"
    >
      <WithStyles(ForwardRef(Button))
        classes={
          Object {
            "root": "reset-button",
          }
        }
        onClick={[Function]}
        variant="contained"
      >
        Reset
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        classes={
          Object {
            "root": "new-button",
          }
        }
        onClick={[Function]}
        variant="contained"
      >
        New Game
      </WithStyles(ForwardRef(Button))>
    </div>
  </div>
</div>
`;
